invoice/invoice.go

```diff
@@ -0,0 +1,966 @@
+// from contains that field. When encoding, only the non-nil fields will be
+// added to the encoded invoice.
+type Invoice struct {
+	// Net specifies what network this Lightning invoice is meant for.
+	Net *chaincfg.Params
+
+	// MilliSat specifies the amount of this invoice in milli-satoshi.
+	// Optional.
+	MilliSat *uint64
+
+	// Timestamp specifies the time this invoice was created.
+	Timestamp uint64 // Mandatory
+
+	// PaymentHash is the payment hash to be used for a payment to this
+	// invoice.
+	PaymentHash []byte
+
+	// PubKey of the target node.
+	// Optional.
+	PubKey *btcec.PublicKey
+
+	// RecoveredPubKey is set to the public key recovered from an encoded
+	// invoice. Will only be non-nil if PubKey is nil. Will be ignored
+	// when encoding.
+	RecoveredPubKey *btcec.PublicKey
+
+	// Description is a short description of the purpose of this invoice.
+	// Optional. Non-nil iff DescriptionHash is nil.
+	Description *string
+
+	// DescriptionHash is the SHA256 hash of a description of the purpose of
+	// this invoice.
+	// Optional. Non-nil iff Description is nil.
+	DescriptionHash []byte
+
+	// Expiry specifies the timespan this invoice will be valid.
+	// Optional. If not set, a default expiry of 60 min will be implied.
+	Expiry *uint64
+
+	// FallbackAddr is an on-chain address that can be used for payment in
+	// case the Lightning payment fails.
+	// Optional.
+	FallbackAddr btcutil.Address
+
+	// RoutingInfo is one or more entries containing extra routing
+	// information for a private route to the target node.
+	// Optional.
+	RoutingInfo []ExtraRoutingInfo
+}
+
+// ExtraRoutingInfo holds the information needed to route a payment along one
+// private channel.
+type ExtraRoutingInfo struct {
+	PubKey       *btcec.PublicKey
+	ShortChanID  uint64
+	Fee          uint64
+	CltvExpDelta uint16
+}
+
+// Amount is a functional option that allows callers of NewInvoice to set the
+// amount in millisatoshis that the Invoice should encode.
+func Amount(milliSat uint64) func(*Invoice) {
+	return func(i *Invoice) {
+		i.MilliSat = &milliSat
+	}
+}
+
+// PubKey is a functional option that allows callers of NewInvoice to explicitly
+// set the pubkey of the Invoice's target node.
+func PubKey(pubKey *btcec.PublicKey) func(*Invoice) {
+	return func(i *Invoice) {
+		i.PubKey = pubKey
+	}
+}
+
+// Description is a functional option that allows callers of NewInvoice to set
+// the payment description of the created Invoice.
+// Note: Must be used if and only if DescriptionHash is not used.
+func Description(description string) func(*Invoice) {
+	return func(i *Invoice) {
+		i.Description = &description
+	}
+}
+
+// DescriptionHash is a functional option that allows callers of NewInvoice to
+// set the payment description hash of the created Invoice.
+// Note: Must be used if and only if Description is not used.
+func DescriptionHash(descriptionHash []byte) func(*Invoice) {
```

Comment:
Should be morphed to be `[32]byte` same goes for when the payment hash itself is specified.